scuttlebutt-mux/src/main/java/net/consensys/cava/scuttlebutt/mux/RPCHandler.java

```diff
@@ -0,0 +1,127 @@
+import java.util.function.Consumer;
+
+public class RPCHandler implements Multiplexer, ClientHandler {
+
+  private final Consumer<Bytes> messageSender;
+  private Map<Integer, CompletableFuture<RPCMessage>> awaitingAsyncResponse = new HashMap<>();
+
+  private Map<Integer, ScuttlebuttStreamHandler> streams = new HashMap<>();
+
+  /**
+   *
+   * @param messageSender sends the request to the server
+   */
+  public RPCHandler(Consumer<Bytes> messageSender, Runnable terminationFn) {
+    this.messageSender = messageSender;
+  }
+
+
+  @Override
+  public CompletableFuture<RPCMessage> makeAsyncRequest(RPCMessage request) {
+    CompletableFuture<RPCMessage> future = new CompletableFuture<>();
+
+    byte rpcFlags = request.rpcFlags();
+    boolean isStream = RPCFlag.Stream.STREAM.isApplied(rpcFlags);
+
+    if (isStream) {
+      future.completeExceptionally(new Exception("Expected async type request, not stream"));
+    } else {
+      int requestNumber = request.requestNumber();
+      awaitingAsyncResponse.put(requestNumber, future);
+
+      Bytes bytes = RPCCodec.encodeRequest(request.body(), requestNumber, rpcFlags);
+
+      messageSender.accept(bytes);
+    }
+
+    return future;
+  }
+
+  @Override
+  public void openStream(RPCMessage request, ScuttlebuttStreamHandler responseSink) {
+    CompletableFuture<RPCMessage> future = new CompletableFuture<>();
+
+    byte rpcFlags = request.rpcFlags();
+    boolean isStream = RPCFlag.Stream.STREAM.isApplied(rpcFlags);
+
+    if (!isStream) {
+      future.completeExceptionally(new Exception("Expected stream type request, not sync type request"));
+    } else {
+      int requestNumber = request.requestNumber();
+      streams.put(requestNumber, responseSink);
+    }
+
+  }
+
+  @Override
+  public void receivedMessage(Bytes message) {
+
+    RPCMessage response = new RPCMessage(message);
+
+    int requestNumber = response.requestNumber() * -1;
+    byte rpcFlags = response.rpcFlags();
+
+    boolean isStream = RPCFlag.Stream.STREAM.isApplied(rpcFlags);
+
+    if (isStream) {
+      ScuttlebuttStreamHandler scuttlebuttStreamHandler = streams.get(requestNumber);
+
+      if (scuttlebuttStreamHandler != null) {
+        scuttlebuttStreamHandler.onMessage(response);
+
+        boolean lastMessageOrError = response.lastMessageOrError();
+
+        if (lastMessageOrError) {
+          streams.remove(requestNumber);
+          scuttlebuttStreamHandler.onStreamEnd();
+        }
+      } else {
+        System.out.println("couldn't find stream handler for RPC response with request number " + requestNumber);
+      }
+
+    } else {
+
+      CompletableFuture<RPCMessage> rpcMessageFuture = awaitingAsyncResponse.get(requestNumber);
+
+      if (rpcMessageFuture != null) {
+
+        rpcMessageFuture.complete(response);
+
+        awaitingAsyncResponse.remove(requestNumber);
+
+      } else {
+        System.out.println("couldn't find future handler for RPC response with request number " + requestNumber);
```

Comment:
don't use System.out, throw errors if you must.

Reply:
Yes, sorry, I'd just posted this temporary code to review the general direction I was going in with it

Reply:
no worries.